pyuvsim/source.py

```diff
@@ -112,17 +114,137 @@ def __init__(self, name, ra, dec, freq, stokes,
         # The coherency is a 2x2 matrix giving electric field correlation in Jy.
         # Multiply by .5 to ensure that Trace sums to I not 2*I
         # Shape = (2,2,Ncomponents)
-        self.coherency_radec = .5 * np.array([[self.stokes[0, :] + self.stokes[1, :],
-                                               self.stokes[2, :] - 1j * self.stokes[3, :]],
-                                              [self.stokes[2, :] + 1j * self.stokes[3, :],
-                                               self.stokes[0, :] - self.stokes[1, :]]])
+        self.coherency_radec = simutils.stokes_to_coherency(self.stokes)
 
         self.time = None
 
         assert np.all([self.Ncomponents == l for l in
                        [self.ra.size, self.dec.size, self.freq.size,
                         self.stokes.shape[1]]]), 'Inconsistent quantity dimensions.'
 
+    def _calc_average_rotation_matrix(self, telescope_location):
+        """
+        Calculate the "average" rotation matrix from RA/Dec to AltAz.
+
+        This gets us close to the right value, then need to calculate a correction
+        for each source separately.
+
+        Parameters
+        ----------
+        telescope_location : astropy EarthLocation object
+            Location of the telescope.
+
+        Returns
+        -------
+        array of floats
+            Rotation matrix that defines the average mapping (RA,Dec) <--> (Alt,Az),
+            shape (3, 3).
+        """
+        # unit vectors to be transformed by astropy
+        x_c = np.array([1., 0, 0])
+        y_c = np.array([0, 1., 0])
+        z_c = np.array([0, 0, 1.])
+
+        # astropy 2 vs 3 use a different keyword name
+        if six.PY2:
+            rep_keyword = 'representation'
+        else:
+            rep_keyword = 'representation_type'
+
+        rep_dict = {}
+        rep_dict[rep_keyword] = 'cartesian'
+        axes_icrs = SkyCoord(x=x_c, y=y_c, z=z_c, obstime=self.time,
+                             location=telescope_location, frame='icrs',
+                             **rep_dict)
+
+        axes_altaz = axes_icrs.transform_to('altaz')
+        setattr(axes_altaz, rep_keyword, 'cartesian')
+
+        ''' This transformation matrix is generally not orthogonal
+            to better than 10^-7, so let's fix that. '''
+
+        from scipy.linalg import orthogonal_procrustes as ortho_procr
```

Comment:
Should this be moved to the top?

Reply:
Yes, I think so since SciPy is listed as a required dependency.